File: src/main/java/site/gachontable/domain/admin/service/GetSeatingsImpl.java
package site.gachontable.domain.admin.service;

import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import site.gachontable.domain.admin.port.out.AdminRepository;
import site.gachontable.domain.admin.exception.AdminNotFoundException;
import site.gachontable.domain.admin.port.in.GetSeatings;
import site.gachontable.presentation.admin.dto.response.SeatingsResponse;
import site.gachontable.infra.security.principal.AuthDetails;
import site.gachontable.domain.pub.domain.Pub;
import site.gachontable.domain.seating.domain.Seating;
import site.gachontable.domain.seating.port.out.SeatingRepository;

import java.time.LocalDateTime;

@Service
@RequiredArgsConstructor
public class GetSeatingsImpl implements GetSeatings {

    private final AdminRepository adminRepository;
    private final SeatingRepository seatingRepository;

    @Transactional(readOnly = true)
    @Override
    public SeatingsResponse execute(AuthDetails authDetails) {
        Pub pub = adminRepository.findById(authDetails.getUuid())
                .orElseThrow(AdminNotFoundException::new)
                .getPub();

        return new SeatingsResponse(
                seatingRepository.findAllByPubAndExitTimeAfterOrderByExitTime(pub, LocalDateTime.now())
                        .stream()
                        .map(Seating::toSeatingResponse)
                        .toList()
        );
    }
}

File: src/main/kotlin/site/gachontable/domain/admin/service/GetSeatingsImpl.kt
package site.gachontable.domain.admin.service

import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import site.gachontable.domain.admin.exception.AdminNotFoundException
import site.gachontable.domain.admin.port.`in`.GetSeatings
import site.gachontable.domain.admin.port.out.AdminRepository
import site.gachontable.domain.pub.domain.Pub
import site.gachontable.domain.seating.domain.Seating
import site.gachontable.domain.seating.port.out.SeatingRepository
import site.gachontable.infra.security.principal.AuthDetails
import site.gachontable.presentation.admin.dto.response.SeatingsResponse
import site.gachontable.presentation.admin.dto.response.SeatingsResponse.SeatingResponse
import java.time.LocalDateTime

@Service
class GetSeatingsImpl(
    private val adminRepository: AdminRepository,
    private val seatingRepository: SeatingRepository,
) : GetSeatings {
    @Transactional(readOnly = true)
    override fun execute(authDetails: AuthDetails): SeatingsResponse {
        val pub: Pub = adminRepository.findById(authDetails.uuid)
            .orElse(throw AdminNotFoundException())
            .pub

        return SeatingsResponse(
            seatingRepository.findAllByPubAndExitTimeAfterOrderByExitTime(pub, LocalDateTime.now())
                .stream()
                .map<SeatingResponse> { seating -> Seating.toSeatingResponse(seating) }
                .toList()
        )
    }
}
